engine/lib/luajit-ffi-gen/readme.md

```diff
@@ -111,9 +111,77 @@ for test:
 cargo expand -p luajit-ffi-gen --test basic_test
 ```
 
+## Supported types
+
+**luajit_ffi_gen** attribute supports different types as input parameters and return type in the impl methods.
+
+Following table shows representation of Rust types in the generated code.
+
+| Rust type                            | C API               | Lua c_defs |
+| ------------------------------------ | ------------------- | ---------- |
+| Immutable reference (&T)             | &T                  | T const*   |
+| Mutable reference (&mut T)           | &mut T              | T *        |
+| Self (in return position)            | Box\<T>             | T *        |
+| Basic and copy types (T)             | T                   | T          |
+| String, str                          | *const libc::c_char | cstr       |
+| Option\<T>                           | *mut T              | T *        |
+| Result\<T, E> (only return position) | T, panic on error   | T          |
+
+### (Mutable) references
+
+In the generated C API references are used as is: **&** or **&mut**. Exceptions are only strings and options.
+
+### Self
+
+Self as input parameter is expected as reference (**&self**) or mutable reference (**&mut self**).
+
+In the return position **Self** is boxed: Box\<T>.
+
+### Basic and copy types
+
+Basic (bool, i32, u64, f32, etc.) and copy types (defined in **COPY_TYPES** map) are sent as is via C API.
+
+### Strings
+
+To avoid additional copy C strings in method parameter position is converted unsafely to Rust string.
+
+In return position Rust string is converted to CStr and copied to a static scoped buffer. Pointer to this buffer is returned via C API.
+
+### User defined types (non copyable)
+
+User types in method parameters position are sent by reference or mutable reference. In the return position type is boxed to transfer ownership to caller (Lua).
+
+**TODO**: should we accept user types in parameters position boxed as well?
+
+### Option
+
+Transferred as **\*mut T** and **None** interpreted as **NULL** pointer.
+
+### Result
+
+Accepted only in the return position. Panics on error.
+
 ---
-### Tip
+## Tip
 
 If there are any compilation errors in the generated C API bindings, developer can copy **#[no_mangle]** code generated by **cargo expand** into the *.rs file and disable **#[luajit_ffi_gen]** attribute.
 
 This way it will be much easier to spot the place of the problem.
+
+## Optimization ideas
+
+If after converting engine to use **luajit_ffi_gen** attribute everywhere compilation time significantly increases, code generation can be optimized in 2 stages.
+
+### Stage 1. Regenerate Lua FFI only if Rust code changes
+
+To avoid unnecessary Lua FFI files regeneration we can calculate hash of the [ImplInfo] structure (actually only name and methods fields) and store it in a file. So before generating Lua FFI file we can check if it's hash changed and do regeneration only if it does.
```

Comment:
```suggestion
To avoid unnecessary Lua FFI files regeneration we can calculate the hash of the [ImplInfo] structure and store it in a file. So before generating the Lua FFI code, we can check to see if it's hash changed first before regenerating.
```